torch/csrc/jit/codegen/cuda/contiguity.cpp

```diff
@@ -8,20 +9,455 @@ namespace jit {
+    return;
+  }
+
+  auto inner_pos = std::distance(active_ids_.begin(), inner_it);
+  auto outer_pos = std::distance(active_ids_.begin(), outer_it);
+
+  // Find inputs in the ordered transforms map
+  const auto inner_ordered_it = consistently_ordered_ids_.find(merge->inner());
+  const auto outer_ordered_it = consistently_ordered_ids_.find(merge->outer());
+
+  bool inner_ordered = inner_ordered_it != consistently_ordered_ids_.end();
+  bool outer_ordered = outer_ordered_it != consistently_ordered_ids_.end();
+
+  // Get root ids of the two inputs
+  const auto inner_root_ids_it = id_to_root_ids_.find(merge->inner());
+  const auto outer_root_ids_it = id_to_root_ids_.find(merge->outer());
+
+  TORCH_INTERNAL_ASSERT(
+      inner_root_ids_it != id_to_root_ids_.end() &&
+          outer_root_ids_it != id_to_root_ids_.end(),
+      "Error replaying transforms in contiguous ID checker.");
+
+  const auto& inner_root_ids = inner_root_ids_it->second;
+  const auto& outer_root_ids = outer_root_ids_it->second;
+
+  // TODO: Concretization may prevent contiguous indexing or vectorization.
+  //  It prevents contiguous indexing if the concretization is within the IDs
+  //  that are used for indexing.
+  //  For vectorization it just means we need to make sure the extents of the
+  //  axes to the right of the broadcast root domain in the contigous merge is
+  //  bigger than the vectorization dimension. And that the tensor buffer
+  //  supports the vector word size (always done).
+  bool inner_is_concretized_bcast = merge->inner()->isBroadcast() &&
+      concrete_info_->isConcretized(merge->inner());
+
+  // Update maps
+  // Find the position inner would have to have to be considered ordered
+  auto pos_after_outer = outer_pos + 1;
+  for (; pos_after_outer < active_ids_.size(); pos_after_outer++) {
+    if (active_ids_[pos_after_outer] == nullptr) {
+      // Can't be considered ordered after a nullptr
+      break;
+    }
+    if (active_ids_[pos_after_outer]->isReduction() ||
+        ((active_ids_[pos_after_outer]->isBroadcast() &&
+          !concrete_info_->isConcretized(active_ids_[pos_after_outer])))) {
+      // Skip reduction or broadcast axes that aren't concretized in the fusion
+      continue;
+    }
+    break;
+  }
+
+  // The output is ordered as long as the inputs were ordered and outer position
+  // is directly left of the inner position.
+  bool out_ordered = inner_ordered && outer_ordered;
+  out_ordered = out_ordered &&
+      // If inner_pos is before outer_pos it's not ordered correctly. If for
+      // some reason it's the same, that would be an error.
+      inner_pos > outer_pos &&
+      // Inner could be a broadcast, so doesn't have to be right on
+      // pos_after_outer as that ID (if it exists) should not be a broadcast.
+      // However, merging over a broadcast should be fine.
+      inner_pos <= pos_after_outer && !inner_is_concretized_bcast;
```

Comment:
Confused here. `pos_after_outer` can be a position where it's not considered ordered, so I thought this statement should have `inner_pos < pos_after_outer`, but `pos_after_outer` can also be the position right next to the outer position and ordered, so `<` wouldn't work in that case. 

Also, it seems odd to have `!inner_is_concretized_bcast`. The concretization and broadcast property are checked for `pos_after_outer`, so it seems redundant.

Reply:
Ah, OK, I think I got it. Since we also check `inner_ordered`, just checking `inner_pos <= pos_after_outer` should work.

I assume having `!inner_is_concretized_bcast` is just to avoid a merge with a concretized broadcast. Added a related comment above.

Reply:
Yeah, that's correct, also agree with `inner_is_concretized_bcast` which I think is a remnant of vectorization depending on contig ids, which it shouldn't exactly.

Reply:
Noted in: https://github.com/csarofeen/pytorch/pull/1990/files#r982796887 I actually don't see a way to move this into ContigIDs easily. Leaving here but also checking outer domain.